File: src/main/java/com/fwdekker/randomness/decimal/DecimalSettingsComponent.java
package com.fwdekker.randomness.decimal;

import com.fwdekker.randomness.JavaHelperKt;
import com.fwdekker.randomness.SettingsComponent;
import com.fwdekker.randomness.ui.ButtonGroupKt;
import com.fwdekker.randomness.ui.JDoubleSpinner;
import com.fwdekker.randomness.ui.JIntSpinner;
import com.fwdekker.randomness.ui.JSpinnerRange;
import com.intellij.openapi.ui.ValidationInfo;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.ButtonGroup;
import javax.swing.JCheckBox;
import javax.swing.JPanel;


/**
 * Component for settings of random decimal generation.
 *
 * @see DecimalSettings
 * @see DecimalSettingsAction
 */
public final class DecimalSettingsComponent extends SettingsComponent<DecimalSettings> {
    private JPanel contentPane;
    private JSpinnerRange valueRange;
    private JDoubleSpinner minValue;
    private JDoubleSpinner maxValue;
    private JIntSpinner decimalCount;
    private JCheckBox showTrailingZeroesCheckBox;
    private ButtonGroup groupingSeparatorGroup;
    private ButtonGroup decimalSeparatorGroup;


    /**
     * Constructs a new {@code DecimalSettingsComponent} that uses the singleton {@code DecimalSettings} instance.
     */
    /* default */ DecimalSettingsComponent() {
        this(DecimalSettings.Companion.getDefault());
    }

    /**
     * Constructs a new {@code DecimalSettingsComponent} that uses the given {@code DecimalSettings} instance.
     *
     * @param settings the settings to manipulate with this component
     */
    /* default */ DecimalSettingsComponent(final @NotNull DecimalSettings settings) {
        super(settings);

        loadSettings();
    }


    @Override
    public JPanel getRootPane() {
        return contentPane;
    }

    /**
     * Initialises custom UI components.
     * <p>
     * This method is called by the scene builder at the start of the constructor.
     */
    private void createUIComponents() {
        minValue = new JDoubleSpinner();
        maxValue = new JDoubleSpinner();
        valueRange = new JSpinnerRange(minValue, maxValue);
        decimalCount = new JIntSpinner(0, 0);

        decimalCount.addChangeListener(event -> {
            final int value = ((JIntSpinner) event.getSource()).getValue();
            showTrailingZeroesCheckBox.setEnabled(value > 0);
        });
    }


    @Override
    public void loadSettings(final @NotNull DecimalSettings settings) {
        minValue.setValue(settings.getMinValue());
        maxValue.setValue(settings.getMaxValue());
        decimalCount.setValue(settings.getDecimalCount());
        showTrailingZeroesCheckBox.setSelected(settings.getShowTrailingZeroes());
        ButtonGroupKt.setValue(groupingSeparatorGroup, settings.getGroupingSeparator());
        ButtonGroupKt.setValue(decimalSeparatorGroup, settings.getDecimalSeparator());
    }

    @Override
    public void saveSettings(final @NotNull DecimalSettings settings) {
        settings.setMinValue(minValue.getValue());
        settings.setMaxValue(maxValue.getValue());
        settings.setDecimalCount(decimalCount.getValue());
        settings.setShowTrailingZeroes(showTrailingZeroesCheckBox.isSelected());
        settings.safeSetGroupingSeparator(ButtonGroupKt.getValue(groupingSeparatorGroup));
        settings.safeSetDecimalSeparator(ButtonGroupKt.getValue(decimalSeparatorGroup));
    }

    @Override
    @Nullable
    public ValidationInfo doValidate() {
        return JavaHelperKt.firstNonNull(
            minValue.validateValue(),
            maxValue.validateValue(),
            valueRange.validateValue(),
            decimalCount.validateValue()
        );
    }
}

File: src/main/kotlin/com/fwdekker/randomness/decimal/DecimalSettingsComponent.kt
package com.fwdekker.randomness.decimal

import com.fwdekker.randomness.SchemeComponent
import com.fwdekker.randomness.decimal.DecimalScheme.Companion.DEFAULT_DECIMAL_SEPARATOR
import com.fwdekker.randomness.decimal.DecimalScheme.Companion.DEFAULT_GROUPING_SEPARATOR
import com.fwdekker.randomness.ui.JDoubleSpinner
import com.fwdekker.randomness.ui.JIntSpinner
import com.fwdekker.randomness.ui.JSpinnerRange
import com.fwdekker.randomness.ui.addChangeListenerTo
import com.fwdekker.randomness.ui.getValue
import com.fwdekker.randomness.ui.setValue
import javax.swing.ButtonGroup
import javax.swing.JCheckBox
import javax.swing.JPanel
import javax.swing.JTextField
import javax.swing.event.ChangeEvent


/**
 * Component for settings of random decimal generation.
 *
 * @param settings the settings to edit in the component
 */
@Suppress("LateinitUsage") // Initialized by scene builder
class DecimalSettingsComponent(settings: DecimalScheme) :
    SchemeComponent<DecimalScheme>(settings) {
    private lateinit var contentPane: JPanel
    private lateinit var valueRange: JSpinnerRange
    private lateinit var minValue: JDoubleSpinner
    private lateinit var maxValue: JDoubleSpinner
    private lateinit var decimalCount: JIntSpinner
    private lateinit var showTrailingZeroesCheckBox: JCheckBox
    private lateinit var groupingSeparatorGroup: ButtonGroup
    private lateinit var decimalSeparatorGroup: ButtonGroup
    private lateinit var prefixInput: JTextField
    private lateinit var suffixInput: JTextField

    override val rootPane get() = contentPane


    init {
        loadScheme()

        decimalCount.addChangeListener { showTrailingZeroesCheckBox.isEnabled = decimalCount.value > 0 }
        decimalCount.changeListeners.forEach { it.stateChanged(ChangeEvent(decimalCount)) }
    }


    /**
     * Initialises custom UI components.
     *
     * This method is called by the scene builder at the start of the constructor.
     */
    @Suppress("UnusedPrivateMember") // Used by scene builder
    private fun createUIComponents() {
        minValue = JDoubleSpinner(description = "minimum value")
        maxValue = JDoubleSpinner(description = "maximum value")
        valueRange = JSpinnerRange(minValue, maxValue, MAX_VALUE_RANGE, name = "value")

        decimalCount = JIntSpinner(0, 0, description = "decimal count")
    }

    override fun loadScheme(scheme: DecimalScheme) {
        minValue.value = scheme.minValue
        maxValue.value = scheme.maxValue
        decimalCount.value = scheme.decimalCount
        showTrailingZeroesCheckBox.isSelected = scheme.showTrailingZeroes
        groupingSeparatorGroup.setValue(scheme.groupingSeparator)
        decimalSeparatorGroup.setValue(scheme.decimalSeparator)
        prefixInput.text = scheme.prefix
        suffixInput.text = scheme.suffix
    }

    override fun saveScheme(scheme: DecimalScheme) {
        scheme.minValue = minValue.value
        scheme.maxValue = maxValue.value
        scheme.decimalCount = decimalCount.value
        scheme.showTrailingZeroes = showTrailingZeroesCheckBox.isSelected
        scheme.groupingSeparator = groupingSeparatorGroup.getValue() ?: DEFAULT_GROUPING_SEPARATOR
        scheme.decimalSeparator = decimalSeparatorGroup.getValue() ?: DEFAULT_DECIMAL_SEPARATOR
        scheme.prefix = prefixInput.text
        scheme.suffix = suffixInput.text
    }

    override fun doValidate() =
        minValue.validateValue()
            ?: maxValue.validateValue()
            ?: valueRange.validateValue()
            ?: decimalCount.validateValue()

    override fun addChangeListener(listener: () -> Unit) =
        addChangeListenerTo(
            minValue, maxValue, decimalCount, showTrailingZeroesCheckBox, groupingSeparatorGroup, decimalSeparatorGroup,
            prefixInput, suffixInput,
            listener = listener
        )

    override fun toUDSDescriptor() = "%Dec[]"


    /**
     * Holds constants.
     */
    companion object {
        /**
         * The maximum difference between the minimum and maximum values that can be generated.
         */
        const val MAX_VALUE_RANGE = 1E53
    }
}
